File: benchmarks/cpp/sorting.cpp
// BENCH_SORT_SIZE=2000 BENCH_SORT_SEED=42
#include <cstdlib>
#include <iostream>

static int rand_at(int i, int seed) {
    long long s = static_cast<long long>(seed) + static_cast<long long>(i) * 1103515245LL;
    if (s < 0) s = -s;
    return static_cast<int>(s % 1000000LL);
}

static long long sort_kernel(int size, int seed) {
    long long comparisons = 0;
    long long checksum = 0;
    for (int i = 0; i < size; ++i) {
        for (int j = i + 1; j < size; ++j) {
            int a = rand_at(i, seed);
            int b = rand_at(j, seed);
            ++comparisons;
            if (a > b) checksum += a - b;
            else checksum += b - a;
        }
    }
    return comparisons + checksum;
}

int main() {
    std::cout << sort_kernel(2000, 42) << '\n';
    return 0;
}

File: benchmarks/rust/src/bin/sorting.rs
fn rand_at(i: i32, seed: i32) -> i32 {
    let mut s = seed as i64 + i as i64 * 1_103_515_245;
    if s < 0 {
        s = -s;
    }
    (s % 1_000_000) as i32
}

fn sort_kernel(size: i32, seed: i32) -> i64 {
    let mut comparisons = 0_i64;
    let mut checksum = 0_i64;
    for i in 0..size {
        for j in (i + 1)..size {
            let a = rand_at(i, seed);
            let b = rand_at(j, seed);
            comparisons += 1;
            if a > b {
                checksum += (a - b) as i64;
            } else {
                checksum += (b - a) as i64;
            }
        }
    }
    comparisons + checksum
}

fn main() {
    println!("{}", sort_kernel(2_000, 42));
}
